### PR TITLE
Use the v2 Identity Service API for lookups (MSC2134 + MSC2140)

### PR DESCRIPTION
**Pull request by anoadragon453**
This is a redo of https://github.com/matrix-org/synapse/pull/5897 but with `id_access_token` accepted.

Implements [MSC2134](https://github.com/matrix-org/matrix-doc/pull/2134) plus Identity Service v2 authentication ala [MSC2140](https://github.com/matrix-org/matrix-doc/pull/2140).

Identity lookup-related functions were also moved from `RoomMemberHandler` to `IdentityHandler`.

### CODE REVIEW

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -669,75 +674,43 @@ def do_3pid_invite(
...
-            raise AuthError(401, "No signature from server %s" % (server_hostname,))
-        for key_name, signature in data["signatures"][server_hostname].items():
-            key_data = yield self.simple_http_client.get_json(
-                "%s%s/_matrix/identity/api/v1/pubkey/%s"
-                % (id_server_scheme, server_hostname, key_name)
+        if invitee:
+            yield self.update_membership(
+                requester, UserID.from_string(invitee), room_id, "invite", txn_id=txn_id
             )
-            if "public_key" not in key_data:
```
Comment (anoadragon453) [inline synapse/handlers/room_member.py]: Note: This looks really weird, but it's just because `verify_any_signature` was moved into identity handler.

File: synapse/util/hash.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+
+import unpaddedbase64
+
+
+def sha256_and_url_safe_base64(input_text):
```
Comment (richvdh) [inline synapse/util/hash.py]: not entirely sure this is worth the effort over inlining it. If you import the functions rather than the modules, it's pretty short:

```
lookup_value = encode_base64(sha256(to_hash.encode()).digest(), urlsafe=True)
```

File: synapse/handlers/identity.py [hunk 3]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+        query_params = {}
+        if id_access_token is not None:
+            query_params["id_access_token"] = id_access_token
+
+        # Check what hashing details are supported by this identity server
+        use_v1 = False
+        hash_details = None
+        try:
+            hash_details = yield self.http_client.get_json(
+                "%s/_matrix/identity/v2/hash_details" % id_server, query_params
```
Comment (richvdh) [inline synapse/handlers/identity.py]: best to use a tuple here for consistency:

```suggestion
                "%s/_matrix/identity/v2/hash_details" % (id_server, ), query_params
```

File: synapse/handlers/identity.py [hunk 4]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+            )
+        except (HttpResponseException, ValueError) as e:
+            # Catch HttpResponseExcept for a non-200 response code
+            # Catch ValueError for non-JSON response body
```
Comment (richvdh) [inline synapse/handlers/identity.py]: if we're doing this, should we check for other exceptions like "failed to connect"?

File: synapse/handlers/identity.py [hunk 5]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+        Args:
+            id_server (str): The server name (including protocol and port, if required)
+                of the identity server to use.
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            id_access_token (str|None): The access token to authenticate to the identity
+                server with
+
+        Returns:
+            str: the matrix ID of the 3pid, or None if it is not recognized.
```
Comment (richvdh) [inline synapse/handlers/identity.py]:
```suggestion
            str|None: the matrix ID of the 3pid, or None if it is not recognized.
```

File: synapse/handlers/identity.py [hunk 6]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+        except (HttpResponseException, ValueError) as e:
+            # Catch HttpResponseExcept for a non-200 response code
+            # Catch ValueError for non-JSON response body
+
+            # Check if this identity server does not know about v2 lookups
+            if e.code == 404:
+                # This is an old identity server that does not yet support v2 lookups
+                use_v1 = True
+            else:
+                logger.warn("Error when looking up hashing details: %s" % (e,))
```
Comment (richvdh) [inline synapse/handlers/identity.py]: .warn is deprecated. Also it interpolates automatically:

```suggestion
                logger.warning("Error when looking up hashing details: %s", e)
```

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/room_member.py]: yeah. I would love it if the move of `_lookup_3pid` (and `_verify_any_signature`) could be in a separate, non-functional, PR, which would help with this sort of thing.

File: changelog.d/5897.feature [hunk 7]
```diff
@@ -0,0 +1 @@
+Switch to the v2 lookup API for 3PID invites.
```
Comment (richvdh) [inline changelog.d/5897.feature]: I suggest mentioning the MSC which this is implementing. (Also worth linking to it from the PR description).

Comment (richvdh) [reply changelog.d/5897.feature]: (seems to be 2134, plus the auth bits from 2140?)

File: synapse/util/hash.py (hunk 2 above)
Comment (richvdh) [reply synapse/util/hash.py]: actually, it's fine. do what you think here.

**Comment by richvdh** [comment]
damn, wrong button

File: synapse/handlers/identity.py [hunk 8]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            hash_details (dict[str, str|list]): A dictionary containing hashing information
+                provided by an identity server.
+
+        Returns:
+            Deferred[str|None]: the matrix ID of the 3pid, or None if it is not recognised.
+        """
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details["algorithms"]
```
Comment (richvdh) [inline synapse/handlers/identity.py]: I feel like we should probably behave better than returning a 500 error if the IS returns a thing which doesn't match the shape we expect.

File: synapse/handlers/identity.py [hunk 9]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+    @defer.inlineCallbacks
+    def _lookup_3pid_v2(
+        self, id_server, id_access_token, medium, address, hash_details
+    ):
+        """Looks up a 3pid in the passed identity server using v2 lookup.
+
+        Args:
+            id_server (str): The server name (including protocol and port, if required)
+                of the identity server to use.
+            id_access_token (str): The access token to authenticate to the identity server with
```
Comment (richvdh) [inline synapse/handlers/identity.py]: do we allow this to be None, or not? currently we can pass in None sometimes...

File: synapse/handlers/identity.py [hunk 10]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+            medium (str): The type of the third party identifier (e.g. "email").
+            address (str): The third party identifier (e.g. "foo@example.com").
+            hash_details (dict[str, str|list]): A dictionary containing hashing information
```
Comment (richvdh) [inline synapse/handlers/identity.py]: we haven't even checked that it is a dict at this point...

File: synapse/handlers/identity.py [hunk 11]
```diff
@@ -282,3 +287,192 @@ def requestMsisdnToken(
...
+                {
+                    "id_access_token": id_access_token,
+                    "addresses": [lookup_value],
+                    "algorithm": lookup_algorithm,
+                    "pepper": lookup_pepper,
+                },
+            )
+        except (HttpResponseException, ValueError) as e:
+            # Catch HttpResponseExcept for a non-200 response code
+            # Catch ValueError for non-JSON response body
```
Comment (richvdh) [inline synapse/handlers/identity.py]: again, what about connection/ssl errors?

File: synapse/handlers/room_member.py [hunk 12]
```diff
@@ -871,11 +842,39 @@ def _ask_id_server_for_third_party_invite(
             "sender_avatar_url": inviter_avatar_url,
         }
 
+        # Add the identity service access token to the JSON body and use the v2
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: I feel like this bit (the switch to /v2/ for these APIs) should probably be a separate PR? It's not mentioned in the newsfile, and it seems like a good candidate to stand alone (I don't think it is required before we land the other half?)

File: synapse/util/hash.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/util/hash.py]: Yeah, I did it because the function name quickly tells you what's going on.

File: changelog.d/5897.feature (hunk 7 above)
Comment (anoadragon453) [reply changelog.d/5897.feature]: Good point, and yes that's correct.

File: synapse/handlers/identity.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/handlers/identity.py]: Hm, I think `ValueError` doesn't have a `code` attribute either, which would cause this code to fail.

Should we drop `ValueError`?

File: synapse/handlers/identity.py (hunk 9 above)
Comment (anoadragon453) [reply synapse/handlers/identity.py]: Theoretically `_lookup_3pid_v2` should never be called if `id_access_token` is None, as `hash_details` is an authenticated endpoint.

File: synapse/handlers/identity.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/handlers/identity.py]: Mm, yeah I'm leaning towards dropping `ValueError` since that's considered a different error from an old IS.

File: synapse/handlers/room_member.py (hunk 12 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: v2 lookup won't work without `id_access_token` support, but could merge it even though it would fail 100% of the time?

File: synapse/handlers/room_member.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: Separated into #5978

File: synapse/handlers/room_member.py (hunk 12 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: Ah I see, you meant the 3PID-invite related things. Yes, that can be moved out.

Moved to https://github.com/matrix-org/synapse/pull/5979

**Comment by richvdh** [comment]
I was kinda hoping you'd do the refactor first, *then* add the new code :/

**Comment by anoadragon453** [comment]
> I was kinda hoping you'd do the refactor first, _then_ add the new code :/

Oh... did this break review workflow?

File: synapse/handlers/room_member.py [hunk 13]
```diff
@@ -681,22 +696,74 @@ def do_3pid_invite(
...
+        # Check what hashing details are supported by this identity server
+        use_v1 = False
+        hash_details = None
+        try:
+            hash_details = yield self.simple_http_client.get_json(
+                "%s%s/_matrix/identity/v2/hash_details" % (id_server_scheme, id_server),
+                query_params,
             )
+            if not isinstance(hash_details, dict):
+                logger.warn(
```
Comment (richvdh) [inline synapse/handlers/room_member.py]:
```suggestion
                logger.warning(
```

**Comment by richvdh** [comment]
it's ok. It just would have been easier to review the new code in the place it's going to land, rather than a temporary location. Basically if you do the refactor first, then each PR stands on its own as a clear improvement. This way round, there's a bit of a backwards step in adding new code to a temporary location. Just something to think about for the future.

File: synapse/handlers/room_member.py [hunk 14]
```diff
@@ -681,22 +696,74 @@ def do_3pid_invite(
...
+                    "Got non-dict object when checking hash details of %s: %s",
+                    id_server,
+                    hash_details,
+                )
+                raise SynapseError(
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: not sure if returning a 500 is the right choice here. Given that the existing code returns None under similar circumstances, shouldn't we do likewise?

(Indeed, I don't think this function has any business deciding what error code will be returned from the API, when we don't even know which endpoint was called.)

File: synapse/handlers/room_member.py [hunk 15]
```diff
@@ -681,22 +696,74 @@ def do_3pid_invite(
...
+                )
+        except Exception as e:
+            # Catch HttpResponseExcept for a non-200 response code
+            # Check if this identity server does not know about v2 lookups
+            if isinstance(e, HttpResponseException) and e.code == 404:
+                # This is an old identity server that does not yet support v2 lookups
+                use_v1 = True
+            else:
+                logger.warn("Error when looking up hashing details: %s" % (e,))
+                raise e
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: again, shouldn't we return None?

**Comment by anoadragon453** [comment]
> This way round, there's a bit of a backwards step in adding new code to a temporary location. 

Yeah, I experienced this backporting some things, but didn't realize that it would be a review issue as well.

Might be time for a new sticky note...

File: synapse/handlers/room_member.py [hunk 16]
```diff
@@ -681,22 +696,74 @@ def do_3pid_invite(
...
+                    500, "Invalid hash details received from identity server"
+                )
+        except Exception as e:
+            # Catch HttpResponseExcept for a non-200 response code
+            # Check if this identity server does not know about v2 lookups
+            if isinstance(e, HttpResponseException) and e.code == 404:
+                # This is an old identity server that does not yet support v2 lookups
+                use_v1 = True
+            else:
+                logger.warn("Error when looking up hashing details: %s" % (e,))
```
Comment (richvdh) [inline synapse/handlers/room_member.py]:
```suggestion
                logger.warning("Error when looking up hashing details: %s", e)
```

File: synapse/handlers/room_member.py [hunk 17]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+    @defer.inlineCallbacks
+    def _lookup_3pid_v2(
+        self, id_server, id_access_token, medium, address, hash_details
+    ):
+        """Looks up a 3pid in the passed identity server using v2 lookup.
+
+        Args:
+            id_server (str): The server name (including port, if required)
+                of the identity server to use.
+            id_access_token (str): The access token to authenticate to the identity server with
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: In *theory* the hash endpoint checks that you gave an access token. In *practice* it might not. I think you should figure out how to handle that case.

File: synapse/handlers/room_member.py [hunk 18]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if not supported_lookup_algorithms or not lookup_pepper:
+            raise SynapseError(
+                500, "Invalid hash details received from identity server: %s, %s"
+            )
+
+        # Check if any of the supported lookup algorithms are present
+        if LookupAlgorithm.SHA256 in supported_lookup_algorithms:
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: what if the `algorithms` aren't an array?

File: synapse/handlers/room_member.py [hunk 19]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+            hash_details (dict[str, str|list]): A dictionary containing hashing information
+                provided by an identity server.
+
+        Returns:
+            Deferred[str|None]: the matrix ID of the 3pid, or None if it is not recognised.
+        """
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if not supported_lookup_algorithms or not lookup_pepper:
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: isn't pepper optional if the algorithm is NONE ?

File: synapse/handlers/room_member.py [hunk 20]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+            lookup_value = "%s %s" % (address, medium)
+
+        else:
+            logger.warn(
+                "None of the provided lookup algorithms of %s are supported: %s",
+                id_server,
+                supported_lookup_algorithms,
+            )
+            raise SynapseError(
+                400,
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: is this specced behaviour? (as opposed to returning None)? if so, can you document this behaviour on the docstring here and on _lookup_3pid?

File: synapse/handlers/room_member.py [hunk 21]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+                "%s%s/_matrix/identity/v2/lookup" % (id_server_scheme, id_server),
+                {
+                    "id_access_token": id_access_token,
+                    "addresses": [lookup_value],
+                    "algorithm": lookup_algorithm,
+                    "pepper": lookup_pepper,
+                },
+            )
+        except Exception as e:
+            logger.warn("Error when performing a v2 3pid lookup: %s" % (e,))
```
Comment (richvdh) [inline synapse/handlers/room_member.py]:
```suggestion
            logger.warning("Error when performing a v2 3pid lookup: %s", e)
```

File: synapse/handlers/room_member.py [hunk 22]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+            raise SynapseError(
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: return None?

File: synapse/handlers/room_member.py [hunk 23]
```diff
@@ -710,17 +777,98 @@ def _lookup_3pid(self, id_server, medium, address):
...
+
     @defer.inlineCallbacks
     def _verify_any_signature(self, data, server_hostname):
         if server_hostname not in data["signatures"]:
             raise AuthError(401, "No signature from server %s" % (server_hostname,))
         for key_name, signature in data["signatures"][server_hostname].items():
             key_data = yield self.simple_http_client.get_json(
-                "%s%s/_matrix/identity/api/v1/pubkey/%s"
-                % (id_server_scheme, server_hostname, key_name)
+                "%s/_matrix/identity/api/v1/pubkey/%s" % (server_hostname, key_name)
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: do we not still need the scheme here?

Comment (anoadragon453) [reply synapse/handlers/room_member.py]: ah, good catch

File: synapse/handlers/room_member.py (hunk 20 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: MSC2134 only mentions an `M_INVALID_PARAM` on attempting to use a lookup algorithm that the identity server doesn't support, but doesn't cover the client not supporting any of the algorithms the identity server does.

I think we just assumed lookup could not happen in this case.

So it might just be easier to have lookup "fail" by returning no results. The client might be a bit confused about why the 3PID lookup failed though.

File: synapse/handlers/room_member.py (hunk 19 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: Nope, it's always required. It's to make implementation easier.

File: synapse/handlers/room_member.py (hunk 18 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: But people on the internet aren't evil :D
<sub>fixed</sub>

File: synapse/handlers/room_member.py (hunk 17 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: I've changed around `_lookup_3pid` to only call `_lookup_3pid_v2` if it gets an `id_access_token`, which should solve this now.

Indeed the original code was not backwards compatible and things would fail if an `id_access_token` parameter was not supplied. That should now be fixed.

File: synapse/handlers/room_member.py [hunk 24]
```diff
@@ -710,9 +777,96 @@ def _lookup_3pid(self, id_server, medium, address):
...
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if (
+            not supported_lookup_algorithms
+            or not isinstance(supported_lookup_algorithms, list)
+            or not lookup_pepper
+            or not isinstance(lookup_pepper, str)
+        ):
+            raise SynapseError(
+                500, "Invalid hash details received from identity server: %s, %s"
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: this is missing the interpolation params

File: synapse/handlers/room_member.py [hunk 25]
```diff
@@ -710,9 +777,96 @@ def _lookup_3pid(self, id_server, medium, address):
...
+        # Extract information from hash_details
+        supported_lookup_algorithms = hash_details.get("algorithms")
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if (
+            not supported_lookup_algorithms
+            or not isinstance(supported_lookup_algorithms, list)
+            or not lookup_pepper
+            or not isinstance(lookup_pepper, str)
+        ):
+            raise SynapseError(
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: there's no value to raising a SynapseError with a 500 error code: it may as well be a regular `Exception`. I'm a bit dubious that this is the correct behaviour though: it's not synapse's fault that the id server is giving incorrect values, so this should be `None` or 421 or 4xx or something

File: synapse/handlers/room_member.py (hunk 20 above)
Comment (richvdh) [reply synapse/handlers/room_member.py]: perhaps you could ask the MSC author to clarify the expected behaviour?

> I think we just assumed lookup could not happen in this case.

I'm not quite sure what you mean by this.

File: synapse/handlers/room_member.py [hunk 26]
```diff
@@ -710,9 +777,96 @@ def _lookup_3pid(self, id_server, medium, address):
...
+            logger.warning("Error when performing a v2 3pid lookup: %s", e)
+            raise SynapseError(
+                500, "Unknown error occurred during identity server lookup"
+            )
+
+        # Check for a mapping from what we looked up to an MXID
+        if "mappings" not in lookup_results or not isinstance(
+            lookup_results["mappings"], dict
+        ):
+            logger.debug("No results from 3pid lookup")
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: maybe info?

File: synapse/handlers/room_member.py [hunk 27]
```diff
@@ -681,9 +696,65 @@ def do_3pid_invite(
...
+            address (str): The third party identifier (e.g. "foo@example.com").
+            id_access_token (str|None): The access token to authenticate to the identity
+                server with
+
+        Returns:
+            str|None: the matrix ID of the 3pid, or None if it is not recognized.
+        """
+        # If an access token is present, add it to the query params of the hash_details request
+        query_params = {}
+        if id_access_token is not None:
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: But the `hash_details` call is how I figured out whether to use v1 or v2 :D

File: synapse/handlers/room_member.py (hunk 25 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: Changed to 421. I think it's best to tell the client when they're using an identity server that's giving them malformed results, such that they can know to change it (since they can do that these days).

File: synapse/handlers/room_member.py (hunk 26 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: If a client implements lookup on every character typed though we'd get a lot of useless logs...

File: synapse/handlers/room_member.py (hunk 20 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: > perhaps you could ask the MSC author to clarify the expected behaviour?

That'd be me.

> > I think we just assumed lookup could not happen in this case.
> 
> I'm not quite sure what you mean by this.

You would need to pick a different identity server.

Comment (anoadragon453) [reply synapse/handlers/room_member.py]: After discussing in the privacy room we're thinking a `500` error since old clients will not know what to do with any special error messages.

File: synapse/handlers/room_member.py (hunk 25 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: Changed to 400 as per privacy-dev discussion.

File: synapse/handlers/room_member.py (hunk 20 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: And now 400 again.

File: synapse/handlers/room_member.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: I see what you meant now. I've moved the `hash_details` call into `_lookup_3pid_v2`

File: synapse/handlers/room_member.py (hunk 26 above)
Comment (anoadragon453) [reply synapse/handlers/room_member.py]: My bad, this will only log if the response from the IS is malformed (it didn't have a "mappings" field), so this is actually quite rare.

Changed it to a `warning()` instead of an `info()` however, as we use `warning()` elsewhere for IS issues.

File: synapse/handlers/room_member.py [hunk 28]
```diff
@@ -793,12 +773,38 @@ def _lookup_3pid_v2(
...
+            hash_details = yield self.simple_http_client.get_json(
+                "%s%s/_matrix/identity/v2/hash_details" % (id_server_scheme, id_server),
+                {"access_token": id_access_token},
+            )
+        except HttpResponseException as e:
+            if e.code == 404:
+                raise
+
+            logger.warning("Error when performing a v2 hash_details request: %s", e)
+            raise SynapseError(
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: why do we not just allow the original exception to propagate? it looks like it will be correctly handled?

File: synapse/handlers/room_member.py [hunk 29]
```diff
@@ -710,9 +759,122 @@ def _lookup_3pid(self, id_server, medium, address):
...
+        lookup_pepper = hash_details.get("lookup_pepper")
+        if (
+            not supported_lookup_algorithms
+            or not isinstance(supported_lookup_algorithms, list)
+            or not lookup_pepper
+            or not isinstance(lookup_pepper, str)
+        ):
+            raise SynapseError(
+                400,
+                "Invalid hash details received from identity server: %s%s, %s"
```
Comment (richvdh) [inline synapse/handlers/room_member.py]:
```suggestion
                "Invalid hash details received from identity server %s%s: %s"
```

File: synapse/handlers/room_member.py [hunk 30]
```diff
@@ -835,7 +843,7 @@ def _lookup_3pid_v2(
                 supported_lookup_algorithms,
             )
             raise SynapseError(
-                400,
+                500,
```
Comment (richvdh) [inline synapse/handlers/room_member.py]: I think we decided this should be a 400 too? In any case, don't use SynapseError for 500s.

**Comment by anoadragon453** [comment]
> tests are failing?

Yeah, tests will fail until https://github.com/matrix-org/sytest/pull/697 is merged, which relies on https://github.com/matrix-org/synapse/pull/6013

File: changelog.d/5897.feature [hunk 31]
```diff
@@ -0,0 +1 @@
+Switch to the v2 lookup API for 3PID invites. Implements MSC2134 plus id_access_token authentication for v2 Identity Service APIs from MSC2140.
```
Comment (richvdh) [inline changelog.d/5897.feature]:
```suggestion
Switch to using the v2 Identity Service `/lookup` API where available, with fallback to v1. (Implements [MSC2134](https://github.com/matrix-org/matrix-doc/pull/2134) plus id_access_token authentication for v2 Identity Service APIs from [MSC2140](https://github.com/matrix-org/matrix-doc/pull/2140)).
```
